server/lyft/gateway/autoplan_builder.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/runatlantis/atlantis/server/events/command"
+	"github.com/runatlantis/atlantis/server/events/metrics"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/events/vcs"
+	"github.com/runatlantis/atlantis/server/logging"
+	"github.com/runatlantis/atlantis/server/recovery"
+	"github.com/uber-go/tally"
+	"strconv"
+)
+
+//go:generate pegomock generate -m --use-experimental-model-gen --package mocks -o mocks/mock_autoplan_validator.go AutoplanValidator
+type AutoplanValidator interface {
+	PullRequestHasTerraformChanges(baseRepo models.Repo, headRepo models.Repo, pull models.PullRequest, user models.User) bool
+}
+
+// AutoplanBuilder handles setting up repo cloning and checking to verify of any terraform files have changed
+type AutoplanBuilder struct {
+	Logger                        logging.SimpleLogging
+	Scope                         tally.Scope
+	VCSClient                     vcs.Client
+	PreWorkflowHooksCommandRunner events.PreWorkflowHooksCommandRunner
+	DisableAutoplan               bool
+	Drainer                       *events.Drainer
+	GlobalCfg                     valid.GlobalCfg
+	// AllowForkPRs controls whether we operate on pull requests from forks.
+	AllowForkPRs bool
+	// AllowForkPRsFlag is the name of the flag that controls fork PR's. We use
+	// this in our error message back to the user on a forked PR so they know
+	// how to enable this functionality.
+	AllowForkPRsFlag string
+	// SilenceForkPRErrors controls whether to comment on Fork PRs when AllowForkPRs = False
+	SilenceForkPRErrors bool
+	// SilenceForkPRErrorsFlag is the name of the flag that controls fork PR's. We use
+	// this in our error message back to the user on a forked PR so they know
+	// how to disable error comment
+	SilenceForkPRErrorsFlag string
+	// SilenceVCSStatusNoPlans is whether autoplan should set commit status if no plans
+	// are found
+	silenceVCSStatusNoPlans bool
+	// SilenceVCSStatusNoPlans is whether any plan should set commit status if no projects
+	// are found
+	silenceVCSStatusNoProjects bool
+	CommitStatusUpdater        events.CommitStatusUpdater
+	PrjCmdBuilder              events.ProjectPlanCommandBuilder
+	PullUpdater                *events.PullUpdater
+}
+
+func (r *AutoplanBuilder) PullRequestHasTerraformChanges(baseRepo models.Repo, headRepo models.Repo, pull models.PullRequest, user models.User) bool {
+	if opStarted := r.Drainer.StartOp(); !opStarted {
+		if commentErr := r.VCSClient.CreateComment(baseRepo, pull.Num, ShutdownComment, command.Plan.String()); commentErr != nil {
+			r.Logger.Log(logging.Error, "unable to comment that Atlantis is shutting down: %s", commentErr)
+		}
+		return false
+	}
+	defer r.Drainer.OpDone()
+
+	log := r.Logger.WithHistory(
+		"repository", baseRepo.FullName,
+		"pull-num", strconv.Itoa(pull.Num),
+	)
+	defer r.logPanics(baseRepo, pull.Num, log)
+
+	scope := r.Scope.SubScope("gateway-autoplan")
```

Comment:
Remember to think about the hierarchy and making things searchable/unified. I think something like this would be cleaner
```
atlantis
  gateway/ <- all gateway metrics would live under here, we can't change the worker path now though since it already exists
    validator/
       valid/
       invalid/
    github/
      event/ <- this path exists under atlantis/ so let's not overwrite them with gateway metrics since we'll be running in hybrid

        
        
  cmd/
     autoplan/...
     comment/...
```